mm2src/coins/solana.rs

```diff
@@ -0,0 +1,599 @@
+        }
+    }
+}
+
+impl From<ParsePubkeyError> for WithdrawError {
+    fn from(e: ParsePubkeyError) -> Self { WithdrawError::InvalidAddress(format!("{:?}", e)) }
+}
+
+impl From<ProgramError> for WithdrawError {
+    fn from(e: ProgramError) -> Self { WithdrawError::InternalError(format!("{:?}", e)) }
+}
+
+#[derive(Debug)]
+pub enum AccountError {
+    NotFundedError(String),
+    ParsePubKeyError(String),
+    ClientError(ClientErrorKind),
+}
+
+impl From<ClientError> for AccountError {
+    fn from(e: ClientError) -> Self { AccountError::ClientError(e.kind) }
+}
+
+impl From<ParsePubkeyError> for AccountError {
+    fn from(e: ParsePubkeyError) -> Self { AccountError::ParsePubKeyError(format!("{:?}", e)) }
+}
+
+impl From<AccountError> for WithdrawError {
+    fn from(e: AccountError) -> Self {
+        match e {
+            AccountError::NotFundedError(_) => WithdrawError::ZeroBalanceToWithdrawMax,
+            AccountError::ParsePubKeyError(err) => WithdrawError::InternalError(err),
+            AccountError::ClientError(e) => WithdrawError::Transport(format!("{:?}", e)),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct SolanaActivationParams {
+    confirmation_commitment: CommitmentLevel,
+    client_url: String,
+}
+
+#[derive(Debug, Display)]
+pub enum SolanaFromLegacyReqErr {
+    InvalidCommitmentLevel(String),
+    InvalidClientParsing(json::Error),
+    ClientNoAvailableNodes(String),
+}
+
+impl SolanaActivationParams {
+    pub fn from_legacy_req(req: &Json) -> Result<Self, MmError<SolanaFromLegacyReqErr>> {
+        let solana_commitment =
+            json::from_value::<String>(req["commitment_level"].clone()).unwrap_or_else(|_| "Finalized".to_string());
+        let solana_client_urls = json::from_value::<Vec<String>>(req["urls"].clone())
+            .map_to_mm(SolanaFromLegacyReqErr::InvalidClientParsing)?;
+        if solana_client_urls.is_empty() {
+            return MmError::err(SolanaFromLegacyReqErr::ClientNoAvailableNodes(
+                "Enable request for SOLANA coin must have at least 1 node URL".to_string(),
+            ));
+        }
+        let commitment_level = match solana_commitment.as_str() {
+            "Finalized" => solana_sdk::commitment_config::CommitmentLevel::Finalized,
+            "Confirmed" => solana_sdk::commitment_config::CommitmentLevel::Confirmed,
+            "Processed" => solana_sdk::commitment_config::CommitmentLevel::Processed,
+            _ => {
+                return MmError::err(SolanaFromLegacyReqErr::InvalidCommitmentLevel(
+                    "Invalid commitment".to_string(),
+                ))
+            },
+        };
+        Ok(SolanaActivationParams {
+            confirmation_commitment: commitment_level,
+            client_url: solana_client_urls[0].clone(),
+        })
+    }
+}
+
+fn generate_keypair_from_slice(priv_key: &[u8]) -> Keypair {
```

Comment:
> Please return `Result` from this fn instead of unwrapping. I also remember that we discussed the following secret generation method - if the wallet passphrase is a valid BIP44 we can use HD derivation. Could you make it this way, please?

Partially solved, now return a result - will add hw support / bip 44 derivation in a separate PR